### PR TITLE
Allow tracking settings to be changed after init to allow for changes during the lifetime of a page, merge in device config to tracking calls, and rework core track logic

### PR DESCRIPTION
**Pull request by rowanbeentje**
These are the base changes required for the app and its single-page model to work with o-tracking; allowing the configuration to be updated without recreating the object means that user etc state can change and be reflected in tracking calls.

Adding context, device, and user from the base config consistently to tracking calls simplifies the code, and means device details can easily be supplied for the first time (we assume this is why Next is currently shoehorning device properties into the user object?).

More than I intended to do but far less than I wanted to do after I started!  Will annotate some parts of the PR...

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -115,33 +115,21 @@ Tracking.prototype.init = function(config) {
 
 	const hasDeclarativeConfig = !!this._getDeclarativeConfigElement();
 
-	if (!(hasDeclarativeConfig || config)) {
```
Comment (rowanbeentje) [inline main.js]: I moved this to far lower - as far as I can tell, if no config is supplied there's no reason not so still set up the class correctly with version and required helpers.

File: src/javascript/core.js [hunk 2]
```diff
@@ -6,69 +6,31 @@
...
- */
 const settings = require('./core/settings');
 const utils = require('./utils');
 
-/**
- * Default properties for sending a tracking request.
- * @type {Object}
- * @return {Object} - The default settings for the component.
- */
-const defaultConfig = function () {
```
Comment (rowanbeentje) [inline src/javascript/core.js]: I significantly refactored this core tracking class to simplify the logic.  I don't see an advantage of having some default config up here and some down in the `track` function.

File: src/javascript/core.js [hunk 3]
```diff
@@ -6,69 +6,31 @@
...
- * Generate and store a new rootID.
- * @param {string} new_id - Optional rootID, if you want to use your own. Otherwise we'll create one for you.
+ * Generate and store a new rootID, used to mark a new root event that
+ * subsequent events will be linked to.
  * @return {string|*} The rootID.
  */
-function setRootID(new_id) {
-	settings.set('root_id', requestID(new_id));
-	return settings.get('root_id');
+function setRootID() {
```
Comment (rowanbeentje) [inline src/javascript/core.js]: While this is technically available externally if something `require()`s in the core object separately, it's not documented and I think it shouldn't be - this is internal behaviour and there's no advantage to allowing it to be customised, and just adds complexity.  It can still be overridden by supplying it as part of the event config.

File: src/javascript/core.js [hunk 4]
```diff
@@ -83,35 +45,48 @@ function track(config, callback) {
 	if (utils.isUndefined(callback)) {
 		callback = function () {};
 	}
-
-	const coreContext = settings.get('config') && settings.get('config').context || {};
```
Comment (rowanbeentje) [inline src/javascript/core.js]: I had troubles tracing out the various directional merges happening across settings context/config context/default config at the top/parameters.  I hope the new layout makes order of overrides clearer.

File: src/javascript/utils.js [hunk 5]
```diff
@@ -49,7 +49,7 @@ function is(variable, type) {
 }
 
 /**
- * Merge objects together. Will remove 'falsy' values.
+ * Merge objects together. Will remove undefined and null values.
```
Comment (rowanbeentje) [inline src/javascript/utils.js]: I've updated this to reflect what the code is actually checking for.  But this isn't quite right either - what it's really doing is ignoring undefined and null values in the object being merged in.  So there's no way to delete an existing key right now, while from this I assumed that using a falsey value would unset the key from the merged object...

File: test/core.test.js [hunk 6]
```diff
@@ -62,11 +71,11 @@ describe('Core', function () {
 			// Context
 			assert.deepEqual(Object.keys(sent_data.context), ["id","root_id","url"]);
 			assert.ok(guid_re.test(sent_data.context.id), "Request ID is invalid. " + sent_data.context.id);
-			assert.equal(sent_data.context.root_id, "root_id");
+			assert.equal(sent_data.context.root_id, root_id);
 			assert.equal(sent_data.context.url, "http://www.ft.com/home/uk");
 
 			// User
-			assert.deepEqual(Object.keys(sent_data.user), ["user_id"]);
+			assert.deepEqual(Object.keys(sent_data.user), ["passport_id","ft_session","user_id"]);
```
Comment (rowanbeentje) [inline test/core.test.js]: I think this was a bug with the old core track implementation.  The documentation says "o-tracking will [...] Automatically pickup ftsession from cookies for you", and I don't think there's an expectation that setting a single user key will unset the defaults, unlike the other arrays.

**Comment by rowanbeentje** [comment]
Although this is technically a major change because it removes the setRootID support from an "external" core singleton, I vote minor change.  Not fussed though.

File: src/javascript/core/user.js [hunk 7]
```diff
@@ -8,6 +8,7 @@ let store;
 const defaultUserConfig = {
 	storage: 'cookie',
 	name: 'spoor-id',
+	nameOverride: 'spoor-id',
 	value: null,
 	domain: (document.URL.match(/^https?\:\/\/([^\/?#]+)(?:[\/?#]|$)/i)[1].indexOf('ft.com') > -1 ? 'ft.com' : null)
```
Comment (JakeChampion) [inline src/javascript/core/user.js]: Should this use `location.hostname` instead of `(document.URL.match(/^https?\:\/\/([^\/?#]+)(?:[\/?#]|$)/i)[1]`?

Comment (JakeChampion) [inline src/javascript/core/user.js]: This could also use `String.prototype.includes` instead of `.indexOf` as it is included in the default polyfill bundle -- https://polyfill.io/v2/docs/features/#String_prototype_includes

File: src/javascript/core.js (hunk 3 above)
Comment (JakeChampion) [inline src/javascript/core.js]: Have you found any code which did consume the core object separately?

File: src/javascript/core/user.js (hunk 7 above)
Comment (rowanbeentje) [inline src/javascript/core/user.js]: Shall I do that in a separate PR?

Agreed on using built-in instead of a regex.  It would also currently match on "microsoft.com"...

Comment (JakeChampion) [inline src/javascript/core/user.js]: Exactly 😂 . Separate PR is good with me 👌

File: src/javascript/core.js (hunk 3 above)
Comment (rowanbeentje) [inline src/javascript/core.js]: Not based on going through https://github.com/search?q=org%3AFinancial-Times+o-tracking+core&type=Code .  Stash string matching with hyphens is terrible so I gave up going through the 1000+ search results there...

Comment (JakeChampion) [inline src/javascript/core.js]: I'm tempted to say _minor_ instead of _major_.

Comment (adambraimbridge) [inline src/javascript/core.js]: Perhaps 'root_id' was originally there to support testing? I'm completely clutching at straws with that guess.

Comment (rowanbeentje) [inline src/javascript/core.js]: Plausible, yes!  I like the way it returns it on generation though - better way of testing it...

**Comment by rowanbeentje** [comment]
@chrisbrownuk be aware that as this gets rolled out, there may be a slight change in user data, but it should always be an improvement (previously if a custom user field was set, the default ft_session etc keys were overwritten rather than added to).
